catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/security/accesscontrol/AccessControlPreQueryPlugin.java

```diff
@@ -0,0 +1,182 @@
+ *   <li>The predicate falls under the logical composite {@link org.opengis.filter.Not} operator.
+ * </ul>
+ */
+public class AccessControlPreQueryPlugin implements PreQueryPlugin {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(AccessControlPreQueryPlugin.class);
+
+  private final FilterBuilder filterBuilder;
+
+  private final SubjectIdentity identity;
+
+  private final AccessControlTags tagSet;
+
+  private final AccessControlSecurityConfiguration configuration;
+
+  public AccessControlPreQueryPlugin(
+      FilterBuilder filterBuilder,
+      SubjectIdentity identity,
+      AccessControlTags tagSet,
+      AccessControlSecurityConfiguration configuration) {
+    this.filterBuilder = filterBuilder;
+    this.identity = identity;
+    this.tagSet = tagSet;
+    this.configuration = configuration;
+  }
+
+  @Override
+  public QueryRequest process(QueryRequest input)
+      throws PluginExecutionException, StopProcessingException {
+    final String subjectIdentifier = getSubjectIdentifier();
+    final Set<String> groups = new HashSet<>(getSubjectGroups());
+    final String groupThatCanSeeEverything = configuration.getSystemUserAttributeValue();
+
+    if (groups.contains(groupThatCanSeeEverything)) {
+      LOGGER.debug(
+          "Will not modify filter; subject ({}) had at least one group [{}] that was exempt [{}]",
+          subjectIdentifier,
+          groups,
+          groupThatCanSeeEverything);
+      return input;
+    }
+
+    final Query query = input.getQuery();
+    final SimplifyingFilterVisitor simpleVisitor = new SimplifyingFilterVisitor();
+    final Filter simplified = (Filter) query.accept(simpleVisitor, null);
+
+    LOGGER.trace("Simplified filter [{}]", simplified);
+
+    final TagAggregationVisitor tagVisitor = new TagAggregationVisitor();
+    simplified.accept(tagVisitor, null);
+
+    final Set<String> discoveredTags = tagVisitor.getTags();
+    if (CollectionUtils.isEmpty(discoveredTags)) {
+      LOGGER.debug(
+          "Will not modify filter; subject's ({}) query [{}] did not completely bind results to tags",
+          subjectIdentifier,
+          query);
+      return input;
+    }
+
+    final Set<String> tagsThatAreAccessControlled = tagSet.getAccessControlledTags();
+    final Set<String> tagsNotAccessControlled =
+        Sets.difference(discoveredTags, tagsThatAreAccessControlled);
+    if (CollectionUtils.isNotEmpty(tagsNotAccessControlled)) {
+      LOGGER.debug(
+          "Will not modify filter; subject's ({}) query [{}] referenced tags that are not in access controlled set [{}]",
+          subjectIdentifier,
+          query,
+          tagsThatAreAccessControlled);
+      return input;
+    }
+
+    final Filter policyBranch = createSecurityPolicySubset(subjectIdentifier, groups);
+    LOGGER.debug("Adding access control policy to the query filter [{}]", policyBranch);
```

Comment:
I thought about that. I had it like that originally, but for testing purposes I wanted the plugin to at least give you one message at `DEBUG` about it's overall decision instead of being silent by default. Going to `TRACE` for that gets really noisy.